src/intrinsics/prepack/global.js

```diff
@@ -277,14 +277,45 @@ export default function(realm: Realm): void {
     );
   }
 
+  function deriveNativeFunctionValueForResidualStatement(
+    name: string,
+    buildNode_: (Array<BabelNodeExpression>) => BabelNodeStatement,
+    numArgs: number
+  ): NativeFunctionValue {
+    return new NativeFunctionValue(realm, "global." + name, name, numArgs, (context, ciArgs) => {
+      invariant(realm.generator !== undefined);
+      realm.generator.emitStatement(ciArgs, buildNode_);
+      return realm.intrinsics.undefined;
+    });
+  }
+
+  // Helper function that specifies a dynamic invariant that cannot be evaluated at prepack time, and needs code to
+  // be injected into the serialized output.
+  global.$DefineOwnProperty("__invariant", {
```

Comment:
I don't have a better name to suggest, but I do feel a little sad about this. In Spec# we had "assert" for compile time invariants that should be proved by the compiler and "assume" for runtime invariants that can be exploited by the compiler. That seems SO much nicer than "invariant" vs "__invariant".

Reply:
Come to think of it, you should push the invariant condition onto the path, because execution cannot continue unless it is true.